pkg/resourcegroup/runaway/manager.go

```diff
@@ -176,11 +178,21 @@ func (rm *Manager) RunawayRecordFlushLoop() {
 			flushRunawayRecords()
 			fired = true
 		case r := <-recordCh:
-			records = append(records, r)
+			key := recordKey{
+				ResourceGroupName: r.ResourceGroupName,
+				SQLDigest:         r.SQLDigest,
+				PlanDigest:        r.PlanDigest,
+				Match:             r.Match,
+			}
+			if _, exists := recordMap[key]; exists {
+				recordMap[key].Repeats++
+			} else {
+				recordMap[key] = r
+			}
 			failpoint.Inject("FastRunawayGC", func() {
 				flushRunawayRecords()
 			})
-			if len(records) >= flushThreshold {
+			if len(recordMap) >= flushThreshold {
```

Comment:
Should we consider the repeats here? For example, if we have only one key but repeat for `flushThreshold` times.

Reply:
Yes, this pr has considered this situation.
https://github.com/pingcap/tidb/blob/6b76e24e450273bf9f6fa3c0ca94c5f9f3d06b2c/pkg/resourcegroup/runaway/manager.go#L187-L189
We only update the repeat field of the value, not adding a new record.

> if we have only one key but repeat for flushThreshold times.

So for this situation, only one record in the map

Reply:
make sense, nice catch!

Reply:
added

Reply:
After discussing, it is unnecessary to add a repeats check.